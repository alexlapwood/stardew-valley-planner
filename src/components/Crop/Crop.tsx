import * as React from "react";

import Sprite from "../Sprite/Sprite";

interface IProps {
  age: number;
  id: string;
  isFlower?: boolean;
  name: string;
  regrow?: number;
  stages: number[];
  x: number;
  y: number;
}

const cropMap = [
  "Amaranth",
  "Ancient Fruit",
  "Artichoke",
  "Beet",
  "Blue Jazz",
  "Blueberry",
  "Bok Choy",
  "Cauliflower",
  "Corn",
  "Cranberry",
  "Eggplant",
  "Fairy Rose",
  "Garlic",
  "Grape",
  "Green Bean",
  "Hops",
  "Hot Pepper",
  "Kale",
  "Melon",
  "Parsnip",
  "Poppy",
  "Potato",
  "Pumpkin",
  "Radish",
  "Red Cabbage",
  "Rhubarb",
  "Starfruit",
  "Strawberry",
  "Summer Spangle",
  "Sunflower",
  "Sweet Gem Berry",
  "Tomato",
  "Tulip",
  "Wheat",
  "Yam"
];

class Crop extends React.Component<IProps> {
  public render() {
    const { age, isFlower, name, regrow, stages, x, y } = this.props;

    const stage = this.calculateStageOfCrop(age, stages, regrow);

    if (stage === undefined) {
      console.log("This should never happen");
      return <div />;
    }

    const spriteIndex = stage + 1;

    return (
      <div
        style={{
          left: `${x * 16}px`,
          position: "absolute",
          top: `${(y - 1) * 16}px`
        }}
      >
        <div style={{ display: "block", position: "relative" }}>
          <Sprite
            height={32}
            src="/images/crops.png"
            width={16}
            x={spriteIndex * 16}
            y={cropMap.indexOf(name) * 32}
          />
          {isFlower &&
            spriteIndex > stages.length && (
              <div style={{ position: "absolute", top: 0 }}>
                <Sprite
                  height={32}
                  src="/images/crops.png"
                  width={16}
                  x={(spriteIndex + 1) * 16}
                  y={cropMap.indexOf(name) * 32}
                />
              </div>
            )}
        </div>
      </div>
    );
  }

  public calculateStageOfCrop(age: number, stages: number[], regrow?: number) {
    let stage = 0;
    let day = 0;

    while (age > day + stages[stage] && stages[stage] !== undefined) {
      day += stages[stage];
      stage += 1;
    }

    if (age > day + 1 && stages[stage] === undefined) {
      day += 1;
      stage += 1;
    }

    if (stage > stages.length) {
      if (regrow !== undefined) {
        const daysSinceLastHarvest = (age - day) % regrow;
        return daysSinceLastHarvest === 0 ? stage - 1 : stage;
      }

      return;
    }

    return stage;
  }
}

export default Crop;
