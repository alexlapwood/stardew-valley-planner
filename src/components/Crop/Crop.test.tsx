import { shallow } from "enzyme";
import * as React from "react";

import Crop from "./Crop";

describe("Crop", () => {
  const defaultProps = {
    age: 0,
    id: "test_crop",
    name: "Test Crop",
    stages: [1, 2, 2, 5]
  };

  it("renders correctly", () => {
    const crop = shallow(<Crop {...defaultProps} />);

    expect(crop.render()).toMatchSnapshot();
  });

  it("renders flower head correctly", () => {
    const crop = shallow(<Crop {...defaultProps} age={11} isFlower={true} />);

    expect(crop.render()).toMatchSnapshot();
  });

  it("renders correctly after it has been removed", () => {
    const crop = shallow(<Crop {...defaultProps} age={28} />);
    expect(crop.render()).toMatchSnapshot();
  });

  it("calculates its growth correctly", () => {
    const crop = shallow(<Crop {...defaultProps} />);
    const cropInstance = crop.instance() as Crop;

    for (let age = 1; age < 28; age += 1) {
      const cropStage = cropInstance.calculateStageOfCrop(
        age,
        cropInstance.props.stages
      );
      switch (age) {
        case 1:
          expect(cropStage).toBe(0);
          break;
        case 2:
        case 3:
          expect(cropStage).toBe(1);
          break;
        case 4:
        case 5:
          expect(cropStage).toBe(2);
          break;
        case 6:
        case 7:
        case 8:
        case 9:
        case 10:
          expect(cropStage).toBe(3);
          break;
        case 11:
          expect(cropStage).toBe(4);
          break;
        default:
          expect(cropStage).toBe(undefined);
      }
    }
  });

  it("calculates its harvesting dates correctly", () => {
    const crop = shallow(<Crop {...defaultProps} regrow={3} />);
    const cropInstance = crop.instance() as Crop;

    for (let age = 12; age < 28; age += 1) {
      const cropStage = cropInstance.calculateStageOfCrop(
        age,
        cropInstance.props.stages,
        cropInstance.props.regrow
      );
      switch (age) {
        case 12:
        case 13:
          expect(cropStage).toBe(5);
          break;
        case 14:
          expect(cropStage).toBe(4);
          break;
      }
    }
  });
});
