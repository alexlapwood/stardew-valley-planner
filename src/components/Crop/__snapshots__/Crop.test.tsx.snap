// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crop renders correctly 1`] = `
<div
  style="left:0px;position:absolute;top:0px"
>
  <div
    style="display:block;position:relative"
  >
    <div
      style="position:relative"
    >
      <div
        class="sprite"
        style="background-image:url(/images/crops.png);background-position:-16px --32px;height:32px;width:16px"
      />
    </div>
  </div>
</div>
`;

exports[`Crop renders correctly after it has been removed 1`] = `<div />`;

exports[`Crop renders flower head correctly 1`] = `
<div
  style="left:0px;position:absolute;top:0px"
>
  <div
    style="display:block;position:relative"
  >
    <div
      style="position:relative"
    >
      <div
        class="sprite"
        style="background-image:url(/images/crops.png);background-position:-80px --32px;height:32px;width:16px"
      />
    </div>
    <div
      style="position:absolute;top:0"
    >
      <div
        style="position:relative"
      >
        <div
          class="sprite"
          style="background-image:url(/images/crops.png);background-position:-96px --32px;height:32px;width:16px"
        />
      </div>
    </div>
  </div>
</div>
`;
